azkaban-common/src/main/java/azkaban/executor/ExecutionFlowDao.java

```diff
@@ -307,10 +315,16 @@ public int selectAndUpdateExecution(final int executorId) throws ExecutorManager
   public static class SelectFromExecutionFlows implements
       ResultSetHandler<List<Integer>> {
 
-    static String SELECT_EXECUTION_FOR_UPDATE = "SELECT exec_id from execution_flows where "
-        + "status = " + Status.PREPARING.getNumVal()
-        + " and executor_id is NULL and flow_data is NOT NULL ORDER BY submit_time ASC LIMIT 1 FOR "
-        + "UPDATE";
+    static String selectExecutionForUpdateQueryString(final int executorId,
```

Comment:
This is creating a new String each time the function is called. Can this return a constant String, either one which would handle use_executor, or one of 2 depending on use_executor?

Reply:
I don’t see how we could avoid creating a string every time because the parameters to the query are not known at compile time. What I can do is to replace the string concatenations with String.format. 
BTW, this query string will likely get even more parameters when we implement the flow priority feature, which is why I still prefer to use an auxiliary method instead of constant strings.

Reply:
You can use "?" for the parameters. Some information at https://docs.oracle.com/javase/tutorial/jdbc/basics/prepared.html. Also examples in uploadFlowFile() and updateChunksInProjectVersions().

Allocating new String each time means that they will need to get garbage collected. Also, It will need to be re-parsed for the query. The earlier link also mentions reusing a PreparedStatement, although the current code doesn't use this.

Some databases do query caching, in which case having the same String can help find a cache hit, but it looks like this was removed in MySQL 8.0: https://mysqlserverteam.com/mysql-8-0-retiring-support-for-the-query-cache/

Reply:
I understand that creating a new string every time means they have to be garbage collected, but that's not worse than using a string constant with "?" in it I believe: a new string has to be created anyways to substitute those ? with the actual values because strings are immutable so because some elements of the query are not known at compile time there's no way to avoid creating a new string every time. The only difference is whether that new string is created by us or by the JDBC driver, but I don't think that's significant difference.

Reply:
If the text of the String does not need to change, then we can create a constant (final static), and return this constant String. This would avoid allocating memory for each call. It's highly unlikely that a String substituting the "?" would be created. The query with the "?" can be executed on the database directly, with the bind variables specified.

Reply:
I changed the query strings to be constant strings using "?" to define parameters. It looks Java delegates parameter resolution to the Mysql connector. Thanks.